### ISSUE TITLE
Regex Support for Column Name "get()" arguments?

### ISSUE THREAD

**Issue by thomasnield** [davidmoten/rxjava-jdbc#34]
I finally got around to playing with rxjava-jdbc and I love it! But I have one idea that may get shot down, but I thought I'd give it a try. 

My current database wrapper I developed internally has getter methods on my `TabularResultSet`, which I've built around the `ResultSet`.

These getter methods are pretty identical to the ones in this library when calling `get()`. You can call `getInt()`, `getString()`,` getBool()` and pass in a String for the column name to get the respective value for. However, my library also allows you to pass a column name expressed as a regex so more than one naming convention can be accepted for a column. While a little loose, this allows multiple data sources to be swapped in easily that have different column conventions. 

So maybe I'm being a little self-focused, but it would be awesome (at least for me) to have multiple column names to be supported in getters via an optional regex feature.

![regex_getter](https://cloud.githubusercontent.com/assets/7420801/9297177/81d4dff2-4460-11e5-9420-0af10eb26a68.png)

What do you guys think? Would anybody else benefit from this or am I a niche?

And thanks again for creating an awesome library, I really look forward to using this in my projects.

**Comment by davidmoten** [comment]
Glad you're finding the library useful. Thanks for the feedback.

Your idea sounds interesting. So at the moment without support in rxjava-jdbc you could get this by doing:

```java
d.select("select from categories where active='Y'"
    .get(rs -> {
        TabularResultSet r = new TabularResultSet(rs);
        return new Category(r.getInt("(CAT(EGORY)?_)?ID"), ...);
    })
    ...
```

Have I understood correctly?

I'm supposing that you'd prefer to instantiate `TabularResultSet` only once because loading the metadata (column names) for every row is a bit wasteful.

Perhaps a plugin arrangement, like a transform that could be specified for every `ResultSet` either for a single `db.select` or as a default transform in `db`.

So you might specify:

```java
db.resultSetTransform(Func1<ResultSet, ? extends ResultSet> transform) {
    ...
}

//example call
db.resultSetTransform(rs-> new TabularResultSet(rs));
```

What do you think?

**Comment by thomasnield** [comment]
Yes, you got the idea. But like you said the metadata and efficiency is tricky. My TabularResultSet requires a single instantiation, accepts the ResultSet as the constructor argument, and it handles the iteration. I even implemented Java 8 streams on it. A little optimization I also do is instead of evaluating a regex for every field for every record, I cache the column index for each found regex and therefore stop doing regex calculations after the first record or so.

You can look at my TabularResultSet interface and implementations here if you like. Not API-optimized but you'll probably get the idea quickly.
https://github.com/thomasnield/tom-commons/tree/master/src/main/java/org/nield/common/data

So I have to think about how it would fit into here... maybe I can poke around and do a PR if I'm able to find a strategically easy place to put this logic.

**Comment by thomasnield** [comment]
Correction, I didn't cache the indexes here. But I did with an implementation at work. Forgot to merge it back to my GitHub repo. I'll see if I can merge tomorrow.

**Comment by davidmoten** [comment]
A contribution like the pattern matching result set wrapper would be great as a standalone class to be used in a resultset transform. Just include that functionality and no other to start with (Streams methods won't be useful in an RxJava context I don't think).

**Comment by thomasnield** [comment]
Okay I will see what I can come up with and file a PR, sounds simple enough.

**Comment by davidmoten** [comment]
The result set transform is released now as 0.6.3 on Maven Central. Let me know if it suits @thomasnield !

**Comment by thomasnield** [comment]
Okay awesome! I'll take it for a spin. I just finished the `RegexMappableResultSet` (we can change the name) and am ready to test it. I'll go ahead and put in a PR. 

Also, should we have an optional parameter that will allow `null` to be returned on a getter instead of throwing an Exception when a field is not found? That might increase the flexibility.
